bundles/org.openhab.binding.verisure/src/main/java/org/openhab/binding/verisure/internal/VerisureSession.java

```diff
@@ -0,0 +1,908 @@
+                        }
+                        verisureInstallations.put(new BigDecimal(owaInstallation.getGiid()), vInst);
+                    } else {
+                        logger.warn("Failed to get alias and/or giid");
+                        return false;
+                    }
+                }
+            }
+        } else {
+            logger.warn("Failed to set session cookie and auth login, HTTP result code: {}", httpResultCode);
+            return false;
+        }
+        return true;
+    }
+
+    private synchronized boolean logIn() {
+        logger.debug("Attempting to log in to mypages.verisure.com");
+        String url = LOGON_SUF;
+        logger.debug("Login URL: {}", url);
+        int httpStatusCode = postVerisureAPI(url, authstring);
+        if (httpStatusCode != HttpStatus.OK_200) {
+            logger.debug("Failed to login, HTTP status code: {}", httpStatusCode);
+            return false;
+        }
+        return true;
+    }
+
+    private void notifyListeners(VerisureThing thing) {
+        for (DeviceStatusListener listener : deviceStatusListeners) {
+            listener.onDeviceStateChanged(thing);
+        }
+    }
+
+    private void notifyListenersIfChanged(VerisureThing thing, VerisureInstallation installation, String deviceId) {
+        String normalizedDeviceId = deviceId.replaceAll("[^a-zA-Z0-9]+", "");
+        thing.setDeviceId(normalizedDeviceId);
+        VerisureThing oldObj = verisureThings.get(normalizedDeviceId);
+        logger.trace("Old thing: {}", oldObj);
+        logger.trace("Updated thing: {}", thing);
+        if (oldObj == null || !oldObj.equals(thing)) {
+            thing.setSiteId(installation.getInstallationId());
+            thing.setSiteName(installation.getInstallationName());
+            verisureThings.put(thing.getDeviceId(), thing);
+            logger.trace("Notify listener of thing {}", thing);
+            notifyListeners(thing);
+        }
+    }
+
+    private void updateStatus() {
+        logger.debug("VerisureSession:updateStatus");
+
+        verisureInstallations.forEach((installationId, installation) -> {
+            configureInstallationInstance(installation.getInstallationId());
+            int httpResultCode = setSessionCookieAuthLogin();
+            if (httpResultCode == HttpStatus.OK_200) {
+                updateAlarmStatus(VerisureAlarms.class, installation);
+                updateSmartLockStatus(VerisureSmartLocks.class, installation);
+                updateMiceDetectionStatus(VerisureMiceDetection.class, installation);
+                updateClimateStatus(VerisureClimates.class, installation);
+                updateDoorWindowStatus(VerisureDoorWindows.class, installation);
+                updateUserPresenceStatus(VerisureUserPresences.class, installation);
+                updateSmartPlugStatus(VerisureSmartPlugs.class, installation);
+                updateBroadbandConnectionStatus(VerisureBroadbandConnections.class, installation);
```

Comment:
Is it possible to move all this updateStatus logic into their respective VerisureThing classes? It would be better to keep related logic closer together. 
It isn't immediately apparent if you could include these methods as part of the VerisureThing interface, so instead what you can do is add a static method in each VerisureThing subclass:
```
public static synchronized void updateStatus(VerisureSession session, VerisureInstallation installation){...}
```
WDYT?

Reply:
The design thought is that all communication with the Verisure API is done in the VerisureSession class. The ThingHandlers for all things uses VerisureSession to update status and send commands. 

The VerisureThing is a JSON -> java implementation for each thing hat VerisureSession uses as a placeholder for the latest status from the API. So it's primary use is to hold thing status from the API.

When VerisureSession has updated a thing and also determined that it's state really differs from previous poll, then it uses a nofify call to the ThingHandler that is a subscriber to state change events.

So there is no direct communication between VerisureSession and the ThingHandlers.

I'm not the original designer of this design and I've not bother to change it. I would say that it would require quit a lot of re-design to change it.